dart/dynamics/Shape.h

```diff
@@ -191,29 +142,19 @@ class Shape : public virtual common::Subject
   /// Instruct this shape to update its data
   virtual void refreshData();
 
-  /// \brief
-  virtual void draw(renderer::RenderInterface* _ri = nullptr,
-                    const Eigen::Vector4d& _color = Eigen::Vector4d::Ones(),
-                    bool _useDefaultColor = true) const = 0;
-
-  /// Pass in true to prevent this shape from being rendered. Pass in false to
-  /// allow it to render again.
-  void setHidden(bool _hide=true);
+  /// Notify that the color (rgba) of this shape has updated
+  virtual void notifyColorUpdate(const Eigen::Vector4d& color);
```

Comment:
For rendering complex geometry its typically better to work with the "flyweight" pattern. That is, the mesh data (visual geometry, color, texture data, shaders, etc.) is shared among all instances of that mesh. This saves memory and allows you to render instances in a batch.

 A visual mesh node could be associated with this data in a few ways. One way is merely to associate it with an asset tag, and have it loaded by assimp. But that precludes being able to edit a visual mesh or procedurally generate it. So, here's how I would do it:

1. Visual Nodes are associated weakly with one or more shared "visual assets".
2. A visual asset could be either an assimp mesh, a procedurally generated internal mesh, or some other kind of asset. The only requirement is that it can be drawn by the viewer.

That way, to procedurally generate a mesh, you could create its visual asset, then associate it with the node. The viewer's responsibility would be looking up the visual assets of all the nodes, and then drawing them (with an overloaded `draw` function for each kind of mesh).